File: C/BinaryTreeLevelOrderTraversal.c
#include<stdio.h>
#include<malloc.h>

typedef struct Tree {
        int data;
        struct Tree* left;
        struct Tree* right;
} Tree;

typedef struct List {
        struct Tree* node;
        struct List* next;
} List;

static int count = 0;

void MakeTree(Tree* root, int leftData, int rightData) {
        if(leftData != -1) {
                root->left = malloc(sizeof(Tree));
                root->left->data = leftData;
        }
        if(rightData != -1) {
                root->right = malloc(sizeof(Tree));
                root->right->data = rightData;
        }
}

void Enqueue(List** q, Tree* root) {
	List* item = malloc(sizeof(List));
	item->node = root;
	item->next = NULL;
	if(!(*q)) {
		*q = item;
	}
	else {
		(*q)->next = item;
		*q = (*q)->next;
	}
	
}

void Dequeue(List** head) {
	*head = (*head)->next;
}

void TraverseLevelOrder(Tree* root) {
	List* q = NULL;
	Enqueue(&q, root);
	List* head = q;
	while(head) {
		printf("%d-> ", head->node->data);
		if(head->node->left)
			Enqueue(&q, head->node->left);
		if(head->node->right)
                        Enqueue(&q, head->node->right);
		Dequeue(&head);
	}
}

void main() {
	Tree* root = malloc(sizeof(Tree));
        root->data =                            	314;
        MakeTree(root,               		6,                  				6);
        MakeTree(root->left, 		271,               561);
	MakeTree(root->left->left, 28,       0);
        MakeTree(root->left->right,         		-1,    3);
	MakeTree(root->left->right->right,                  17,   -1);
        MakeTree(root->right,                           			2,  			271);
	MakeTree(root->right->left,                                        -1,       1);
	MakeTree(root->right->left->right,                                      401,     257);
	MakeTree(root->right->left->right->left,                             -1,     641);
        MakeTree(root->right->right,                                                   		-1,              28);
	
	TraverseLevelOrder(root);
	printf("\n");
}

File: LeetCode/Medium/Binary-Tree-Level-Order-Traversal.cs
/*
https://leetcode.com/problems/binary-tree-level-order-traversal/

Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).

 

Example 1:


Input: root = [3,9,20,null,null,15,7]
Output: [[3],[9,20],[15,7]]
Example 2:

Input: root = [1]
Output: [[1]]
Example 3:

Input: root = []
Output: []
 

Constraints:

The number of nodes in the tree is in the range [0, 2000].
-1000 <= Node.val <= 1000
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<IList<int>> LevelOrder(TreeNode root) {
        if(root == null)
        {
            return new List<IList<int>>();
        }
        
        var result = new List<IList<int>>();
        int count = 0;
        int nextCount = 0;
        var q = new Queue<TreeNode>();
        q.Enqueue(root);
        nextCount++;
        while(q.Count > 0)
        {
            count = nextCount;
            nextCount = 0;
            var l = new List<int>();
            while(count > 0)
            {
                var node = q.Dequeue();
                if(node.left != null)
                {
                    q.Enqueue(node.left);
                    nextCount++;
                }
                if(node.right != null)
                {
                    q.Enqueue(node.right);
                    nextCount++;
                }
                
                l.Add(node.val);
                count--;
            }
            
            if(l.Count > 0)
            {
                result.Add(l);
            }
        }
        
        return result;
        
    }
}